ballerina/listener_utils.bal

```diff
@@ -102,6 +106,176 @@ isolated function addDefaultDirectives(__Schema schema) {
+            };
+            return file;
+        }
+        return byteStream;
+    }
+    return mediaType;
+}
+
+isolated function getFileUploadValues(map<FileUpload> fileInfo, map<json> pathMap, map<json> variables) returns map<FileUpload|FileUpload[]>|string {
+    map<FileUpload> files = {}; //map of file path and file value
+    if pathMap.length() > 0 {
+        foreach var item in pathMap.entries() {
+            if item[1] is json[] {
+                json[] value = <json[]>item[1];
+                if value.length() > 0 {
+                    foreach json path in value {
+                        if path is string {
+                            if checkAndReplaceVariablePosition(variables, path) {
+                                string key = item[0];
+                                if fileInfo.hasKey(key) {
+                                    files[path] = <FileUpload> fileInfo[key];
+                                } else {
+                                    return "File content is missing in multipart request";
+                                }
+                            } else {
+                                return "Undefined variable found in multipart request `map`";
+                            }
+                        } else {
+                            return "Invalid type for multipart request field ‘map’ value";
+                        }
+                    }
+                } else {
+                    return "Missing multipart request field ‘map’ values";
+                }
+            } else {
+                return "Invalid type for multipart request field ‘map’";
+            }
+        }
+    } else {
+        return "Missing multipart request field ‘map’";
+    }
+    return createFileUploadInfoMap(files, variables);
+}
+
+isolated function checkAndReplaceVariablePosition(map<json> variables, string name) returns boolean {
+    //replace variable's null value with file path
+    string varName = name;
+    if name.includes(".") {
+        varName = name.substring(0, <int> name.indexOf("."));
+        string indexPart = name.substring((<int> name.indexOf(".") + 1), name.length());
+        int|error index = 'int:fromString(indexPart);
+        if variables.hasKey(varName) && variables.get(varName) is json[] && index is int {
+            json[] arrayValue = <json[]> variables.get(varName);
+            if index < arrayValue.length() {
+                arrayValue[index] = name;
+                variables[varName] = arrayValue;
+                return true;
+            }
+            return false;
+        }
+        return false;
+    }
+    return variables.hasKey(varName);
+}
+
+isolated function createFileUploadInfoMap(map<FileUpload> files, map<json> variables) returns map<FileUpload|FileUpload[]>|string {
+    map<FileUpload|FileUpload[]> fileInfo = {};
+    foreach var entry in variables.entries() {
+        string key = entry[0];
+        json value = entry[1];
```

Comment:
A better way to do this is to iterate through all the keys.

```ballerina
foreach string key in variables.keys() {
    json value = variables.get(key);
    // ...
}
```

Using `entries()`, and `entry[0]` and `entry[1]` feels like an anti-pattern.